core/trino-main/src/main/java/io/trino/execution/scheduler/NodeRequirements.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.connector.CatalogName;
+import io.trino.spi.HostAddress;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class NodeRequirements
+{
+    private final Optional<CatalogName> catalogName;
+    private final Optional<Set<HostAddress>> addresses;
```

Comment:
What does it mean for the set of addresses to be empty? Wouldn't that be semantically the same as the Optional being empty?

Reply:
`Optional.empty()` indicates no requirement. `Optional.of(emptySet())` indicates that there's a requirement that no existing node can satisfy (that should never happen in practice). However it makes the distinction between "no requirement" and "some requirement" more explicit. I don't have a strong preference here, we can drop the Optional.